cirq/work/collector.py

```diff
@@ -0,0 +1,164 @@
+    def next_job(self
+                ) -> Union[None, CircuitSampleJob, Iterable[CircuitSampleJob]]:
+        """Called by driving code when more sampling can be started.
+
+        Returns:
+            A CircuitSampleJob describing the circuit to sample, how many
+            samples to take, and a key value that can be used in the
+            `on_job_result` method to recognize which job this is.
+        """
+
+    @abc.abstractmethod
+    def on_job_result(self, job: CircuitSampleJob,
+                      result: study.TrialResult) -> None:
+        """Called by driving code when sample results have become available.
+
+        The results should be incorporated into the collector's state.
+        """
+
+
+async def async_collect_samples(collector: SampleCollector,
+                                sampler: sampler.Sampler,
+                                *,
+                                concurrency: int = 2,
+                                max_total_samples: Optional[int] = None
+                               ) -> None:
+    """Concurrently collects samples from a simulator or hardware sampler.
+
+    Examples:
+
+        ```
+        collector = cirq.PauliStringCollector(...)
+        await cirq.async_collect_samples(
+            sampler=cirq.Simulator(),
+            collector=collector,
+            concurrency=3)
+        print(collector.estimated_energy())
+        ```
+
+    See Also:
+
+        Python 3 documentation "Coroutines and Tasks"
+        https://docs.python.org/3/library/asyncio-task.html
+
+    Args:
+        collector: Determines which circuits to sample next, and processes the
+            resulting samples. This object will be mutated.
+        sampler: The simulator or hardware to take samples from.
+        concurrency: Desired number of sampling jobs to have in flight at any
+            given time.
+        max_total_samples: Optional limit on the maximum number of samples to
+            collect.
+
+    Returns:
+        The collector's result after all desired samples have been collected.
+    """
+    pool = work_pool.CompletionOrderedAsyncWorkPool()
+    remaining_samples = (np.infty
+                         if max_total_samples is None else max_total_samples)
+
+    async def _start_async_job(job):
+        return job, await sampler.async_sample(job.circuit,
+                                               repetitions=job.repetitions)
+
+    # Keep dispatching and processing work.
+    while True:
+        # Fill up the work pool.
+        while (remaining_samples > 0 and pool.num_uncollected < concurrency):
+            new_jobs = _flatten_jobs(collector.next_job())
+
+            # If no jobs were given, stop asking until something completes.
+            if not new_jobs:
+                break
+
+            # Start new sampling jobs.
+            for new_job in new_jobs:
+                remaining_samples -= new_job.repetitions
+                pool.include_work(_start_async_job(new_job))
```

Comment:
Note that this could exceed the specified `concurrency` limit if a collector returns a large chunk of jobs from `next_job`. I think it would be better to include backpressure in the work pool; you would pass the concurrency to the work pool on creation and then here you'd do something like
```python
await pool.ready()
pool.include_work(_start_async_job(new_job))
```
where `pool.ready()` would complete only when the number of outstanding work items drops below the specified limit. Of course, in that case you'd need to run two separate coroutines to submit work to the pool and send results back to the collector as they complete, but I think that would be conceptually better than having a single loop that does both jobs anyway.

Reply:
I fixed this by just keeping a queue of work.

I agree that it would be ideal for the pool to have a method like `await pool.active_jobs_drops_below(x)`.

Reply:
Tried. It has slightly wrong semantics (as discussed).